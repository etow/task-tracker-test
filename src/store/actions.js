import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
