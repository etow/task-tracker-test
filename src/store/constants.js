

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
