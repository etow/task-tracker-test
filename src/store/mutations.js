import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    state.tasks[task.category][targetIndex] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  }
}
