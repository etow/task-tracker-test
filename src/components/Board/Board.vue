<template>
  <div class="board" v-loading="loading">
    <BoardColumn
      v-for="category in Object.keys(tasks)"
      :key="category"
      :category="category"
      :tasks="tasks[category]" />
  </div>
</template>

<script>
import BoardColumn from '../BoardColumn/BoardColumn.vue';
import { FETCH_TASKS_CATEGORIES, FETCH_TASKS, ROLLBACK_STATE } from '../../store/constants';
import { ElMessage } from 'element-plus';

export default {
  name: "Board",
  data() {
    return {
      loading: false,
    }
  },
  components: {
    BoardColumn,
  },
  computed: {
    tasks() {
      return this.$store.state.tasks;
    }
  },
  created() {
    this.loading = true;
    this.$store.dispatch(FETCH_TASKS_CATEGORIES)
      .then(() => this.$store.dispatch(FETCH_TASKS))
      .catch((err) => {
        console.error(err.message);
        ElMessage({
          message: 'Something went wrong while fetching tasks, please try again',
          type: 'error',
          duration: 5000,
        });
        this.$store.dispatch(ROLLBACK_STATE);
      })
      .finally(() => this.loading = false);
  },
};
</script>

<style lang="scss" scoped>
.board {
  display: flex;
  height: 100%;
  background: linear-gradient(-45deg, rgb(245, 247, 248), rgb(237, 241, 242) 100%);
}
</style>
