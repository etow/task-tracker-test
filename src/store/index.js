import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
