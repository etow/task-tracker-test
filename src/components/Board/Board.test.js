import { shallowMount, flushPromises } from '@vue/test-utils';
import { ElMessage } from 'element-plus';
import Board from './Board.vue';
import { createStore } from 'vuex';
import { FETCH_TASKS_CATEGORIES, FETCH_TASKS } from '../../store/constants';

let loadingValue = true;

jest.mock('element-plus', () => ({
  ElMessage: jest.fn(),
  ElLoading: { service: jest.fn(() => ({ close: jest.fn(() => loadingValue = false) })) },
}));

describe('Board.vue', () => {
  it('renders correctly',  async () => {
    // mock store
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('fetch tasks and categories',  async () => {
    // mock store
    const [store, { actions }] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    await flushPromises();
    expect(actions[FETCH_TASKS_CATEGORIES]).toHaveBeenCalled();
    expect(actions[FETCH_TASKS]).toHaveBeenCalled();
    await flushPromises();
    expect(loadingValue).toBe(false);
  });

  it('handles error if fetch fails',  async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const [store, { actions }] = mockStore({ fetch_tasks_categories: jest.fn(() => Promise.reject('error'))});

    createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    expect(actions[FETCH_TASKS_CATEGORIES]).toHaveBeenCalled();
    await flushPromises();

    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while fetching tasks, please try again',
      type: 'error',
      duration: 5000,
    });
    console.error.mockRestore();
  });
});

// create wrapper helper
const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    global: {
      directives: {
        loading() {},
      },
    },
  };

  decorator(dataToMount);
  return shallowMount(Board, dataToMount);
};

// mock store helper
function mockStore({ fetch_tasks_categories, fetch_tasks } = {}) {
  const state = {
    tasks: {
      pending: [],
      inprogress: [],
      completed: [],
    }
  };

  const actions = {
    [FETCH_TASKS_CATEGORIES]: fetch_tasks_categories || jest.fn(() => Promise.resolve()),
    [FETCH_TASKS]: fetch_tasks || jest.fn(() => Promise.resolve()),
  }

  const store = createStore({
    state,
    actions,
  });

  return [store, { actions }];
}
