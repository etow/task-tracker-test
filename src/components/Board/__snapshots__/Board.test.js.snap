// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board.vue renders correctly 1`] = `
<div class="board">
  <board-column-stub category="pending" tasks=""></board-column-stub>
  <board-column-stub category="inprogress" tasks=""></board-column-stub>
  <board-column-stub category="completed" tasks=""></board-column-stub>
</div>
`;
